internal/consumer/worker.go

```diff
@@ -0,0 +1,25 @@
+package consumer
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+func worker(ctx context.Context, idChan chan int, itemProvider ItemProvider, itemRepository ItemRepository, wg sync.WaitGroup) {
+	for id := range idChan {
+		item, err := itemProvider.GetItem(id)
+		if err != nil {
+			log.Print(errors.Wrap(err, "worker: "))
+		} else if !item.Dead && !item.Deleted {
+			_, err := itemRepository.SaveItem(ctx, *item)
+			if err != nil {
+				log.Print(errors.Wrap(err, "worker: "))
+			}
+		}
+	}
+	wg.Done()
```

Comment:
In Go you can use `defer wg.Done()` which is preferred as it is guaranteed to be called when the function returns.

Imagine if you increased the length of this function and then after a few weeks came back to it and slapped a random `return` in the `for` loop. In that instance your `worker` function would return and `wg.Done()` would never be called. Subsequently, you [wg.Wait](https://github.com/JakeHumphries/gymshark-hacker-news/pull/1/files?authenticity_token=rjIYMsk5UfnEkboE6xYvbbY6Zk6%2FWIhjO%2BHympp8f5XPMzorVbirnPQTxby8sge6U3kzlCBjnPBitt3ZN%2Bb8eA%3D%3D&file-filters%5B%5D=.example&file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=dotfile#diff-6c00c19a3da7d27b57a003f96a6d9c1925cee5aa1ea535071f57b4ae3ee227f7R40) would never complete and the program will hang.

Essentially, if you need to clean up after yourself then stick it in a defer to make sure it happens.

For a better explanation than I gave 😄  https://go.dev/blog/defer-panic-and-recover

Reply:
good shout, thanks.